util/internal/integrationtest/orm/src/main/java/org/hibernate/search/util/impl/integrationtest/orm/SimpleSessionFactoryBuilder.java

```diff
@@ -28,8 +28,8 @@
 	private final List<Consumer<MetadataSources>> metadataSourcesContributors = new ArrayList<>();
 	private final List<Consumer<SessionFactoryBuilder>> sessionFactoryBuilderContributors = new ArrayList<>();
 
-	public SimpleSessionFactoryBuilder setTcclLookupPrecedence(TcclLookupPrecedence tcclLookupPrecedence) {
-		return onBootstraServiceRegistryBuilder( builder -> builder.applyTcclLookupPrecedence( tcclLookupPrecedence ) );
+	public SimpleSessionFactoryBuilder setTcclLookupPrecedenceBefore() {
+		return onBootstraServiceRegistryBuilder( builder -> builder.applyClassLoader( new TcclLookupPrecedenceClassLoader() ) );
```

Comment:
I'm really not sure about this.

You're implementing your own classloader just to change the TCCL precedence in Hibernate ORM, which is a feature that is already exposed through SPIs through `org.hibernate.boot.registry.BootstrapServiceRegistryBuilder#applyTcclLookupPrecedence`. There's a fair amount of risk involved in deviating too much from the what our users' environment will be: our tests may end up not being representative.

Yes, `TcclLookupPrecedence` itself is marked as internal, but then it's exposed in `org.hibernate.boot.registry.BootstrapServiceRegistryBuilder` which is clearly an SPI. So in this case an implementation type is leaking through Hibernate ORM's SPI. The problem is in ORM, not in our code.

If you want to limit the usage of `TcclLookupPrecedence`, I'd advise to apply your changes up to this method, and give up here, calling `applyTcclLookupPrecedence( TcclLookupPrecedence.BEFORE )` directly.

If Hibernate ORM breaks this SPI, well, we can always fall back to your solution. It's not a big deal, as this is test code and a compatibility break will not affect our users.

Reply:
Is a SPI method that has an internal parameter a little strange? Anyway let's limit as much as possible the internal exposition.

Reply:
As the comment says... Actually it was my first intention.

Reply:
> Is a SPI method that has an internal parameter a little strange

Yes I totally agree it's strange. But for now we'll have to deal with it.